api/src/main/java/org/openmrs/api/db/hibernate/HibernateProviderDAO.java

```diff
@@ -279,6 +279,16 @@ public ProviderAttributeType getProviderAttributeType(Integer providerAttributeT
 		return (ProviderAttributeType) getSession().get(ProviderAttributeType.class, providerAttributeTypeId);
 	}
 	
+	/**
+	 * @see org.openmrs.api.db.ProviderDAO#getProviderAttributeTypeByName(String)
+	 */
+	@Override
+	public ProviderAttributeType getProviderAttributeTypeByName(String name) {
+		Criteria criteria = getSession().createCriteria(ProviderAttributeType.class);
+		criteria.add(Restrictions.eq("name", name));
+		return (ProviderAttributeType) criteria.uniqueResult();
```

Comment:
What happens when we have more than one provider attribute type with the same name?

Reply:
@dkayiwa should i throw an exception or is having same names for the providerattributetype an acceptable occurance? or should i Set up a unique constraint on the "name" column of the database table storing the provider attribute types. This would prevent the insertion of duplicate names in the first place.

check out the next commit for my thought process on your question :smiling_face_with_tear:

Reply:
What do you think would be the best option?

Reply:
i think what i have done in the latest commit to retrieve a list of ProviderAttributeType objects from the database where the "name" attribute matches the provided name does the trick.

Reply:
And what happens at this code when we have more than one returned?

Reply:
And what if it is the one at index 1 that the caller wanted?

Reply:
i am blocked i guess! at this level i am thinking of modifying the method in the service class to something like `ProviderAttributeType getProviderAttributeTypeByNameAndIndex(String name, int index)` that uses `getProviderAttributeTypesByName(name)` to get the list and then iterate through with something like     
```
if (index >= 0 && index < attributeTypes.size()) {
        return attributeTypes.get(index);
    } else {
        return null;  // Or throw an exception to indicate an invalid index.
    }
```

Reply:
Why don't you do like the rest of the openmrs objects such as person attribute types?

Reply:
This is something we're somewhat inconsistent on. For locations, where we ensure that the attribute type names are unique on creation will actually throw a Hibernate `NonUniqueResultException`, because we use `uniqueResult()`; person attribute types will return result 0 regardless of the number of responses returned. Programs actually throw a `ProgramNameDuplicatedException`.

Reply:
Done!

Reply:
Did you also ensure that one is not able to create duplicate provider attribute types as it is for person attribute types?

Reply:
The prevention of duplication is achieved through the use of the `@Column` annotation on the `getName()` method as seen below. I did this because I could not directly make the 'name' column in BaseOpenmrsdata unique since it is extended by other classes that might not need the uniqueness of name

```java
@Column(name = "name", nullable = false, length = 255, unique = true)
@Field
public String getName() {
    return super.getName();
}
```

The `unique = true` attribute within the `@Column` annotation indicates that the values in the `name` column should be unique which means that each value in the `name` column must be distinct, and any attempt to insert or update a row with a duplicate name value will result in a constraint violation.

Reply:
@mherman22 did you test this and confirm that you are actually not able to create two provider attribute types with the same name?

Reply:
my latest commit addresses your question @dkayiwa

Reply:
well apparently the webapp still allows their creation, i need to figure out why

cc: @dkayiwa